filebeat/input/filestream/input.go

```diff
@@ -168,7 +201,140 @@ func (inp *filestream) Run(
+		case errors.Is(err, ErrInactive) && inp.deleterConfig.OnClose.Inactive:
+			log.Debugf(
+				"'%s' will be removed because 'delete.on_close.inactive' is set",
+				fs.newPath,
+			)
+			deleteFile = true
+		}
+
+		if deleteFile {
+			if err := inp.deleteFile(ctx, log, cursor, fs); err != nil {
+				return fmt.Errorf("cannot remove file '%s': %w", fs.newPath, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (inp *filestream) deleteFile(
+	ctx input.Context,
+	logger *logp.Logger,
+	cursor loginp.Cursor,
+	fs fileSource,
+) error {
+	// Wait until the resource is finished.
+	// The resource is finished when all events are acknowledged by the output.
+	// If it is not finished, this usually means one of two things:
+	//   - The output is down
+	//   - Filebeat is experiencing back pressure
+	// Either way, we wait until all events have been published.
+	if !cursor.Finished() {
+		logger.Debugf(
+			"not all events from '%s' have been published, "+
+				"closing harvester",
+			fs.newPath)
+		// normal operation of the harvester, the file was closed.
+		return nil
+	}
+	logger.Infof(
+		"all events from '%s' have been published, waiting for %s grace period",
+		fs.newPath, inp.deleterConfig.GracePeriod.String())
+
+	graceTimer := time.NewTimer(inp.deleterConfig.GracePeriod)
+	// Wait for the grace period or for the context to be cancelled
+	select {
+	case <-ctx.Cancelation.Done():
+		return ctx.Cancelation.Err()
+	case <-graceTimer.C:
+	}
+
+	// Check if file is still at EOF
+	// We know all events have been published because cursor.Finished is true,
+	// so we can get the offset from the cursor and compare it with the file size.
+	st := state{}
+	if err := cursor.Unpack(&st); err != nil {
+		return fmt.Errorf("cannot unpack cursor from '%s' to read offset: %w",
+			fs.newPath,
+			err)
+	}
+
+	stat, err := os.Stat(fs.newPath)
+	if err != nil {
+		return fmt.Errorf("cannot stat '%s': %w", fs.newPath, err)
+	}
+
+	// The file has been written to, close the harvester so the filewatcher
+	// can start a new one
+	if stat.Size() != st.Offset {
+		logger.Debugf("'%s' was updated, won't remove. Closing harvester", fs.newPath)
+		return nil
+	}
+
+	if err := os.Remove(fs.newPath); err != nil {
+		// The first try at removing the file failed,
+		// retry with a constant backoff
+		lastTry := time.Now()
+		lastErr := err
+
+		maxWait := 10 * time.Second
+		ticker := time.NewTicker(2 * time.Second)
+		defer ticker.Stop()
+
+		retries := 0
+		for maxWait >= 0 {
+			logger.Errorf(
+				"could not remove '%s', retrying in 2s. Error: %s",
+				fs.newPath,
+				err,
+			)
+
+			select {
+			case <-ctx.Cancelation.Done():
+				return ctx.Cancelation.Err()
+
+			case <-ticker.C:
+				retries++
+				lastErr = os.Remove(fs.newPath)
+				if lastErr == nil {
+					logger.Infof("'%s' removed", fs.newPath)
+					return nil
+				}
+				// Calculate the wait using monotonic clock
+				now := time.Now()
+				maxWait -= now.Sub(lastTry)
+				lastTry = now
```

Comment:
Can't you just count the number of times ticker has fired since the wait is constant?